runtime/runtime/src/lib.rs

```diff
@@ -478,15 +479,45 @@ impl Runtime {
                     apply_state,
                     deploy_global_contract,
                     &mut result,
+                    stats,
                 );
             }
             Action::UseGlobalContract(use_global_contract) => {
                 let account = account.as_mut().expect(EXPECT_ACCOUNT_EXISTS);
-                action_use_global_contract(state_update, account, use_global_contract)?;
+                action_use_global_contract(
+                    state_update,
+                    account_id,
+                    account,
+                    use_global_contract,
+                    apply_state.current_protocol_version,
+                    &mut result,
+                )?;
             }
             Action::FunctionCall(function_call) => {
                 let account = account.as_mut().expect(EXPECT_ACCOUNT_EXISTS);
-                let code_hash = account.local_contract_hash().unwrap_or_default();
+                let account_contract = account.contract();
+
+                let code_hash = match account_contract.as_ref() {
+                    AccountContract::None => CryptoHash::default(),
+                    AccountContract::Local(code_hash) | AccountContract::Global(code_hash) => {
+                        *code_hash
+                    }
+                    AccountContract::GlobalByAccount(account_id) => {
+                        let identifier = GlobalContractIdentifier::AccountId(account_id.clone());
+                        let key = TrieKey::GlobalContractCode { identifier: identifier.into() };
+                        let value_ref = state_update
+                            .get_ref(&key, KeyLookupMode::FlatStorage)?
+                            .ok_or_else(|| {
+                                let TrieKey::GlobalContractCode { identifier } = key else {
+                                    unreachable!()
+                                };
+                                RuntimeError::GlobalContractError(
```

Comment:
we should never reach this under normal circumstances since after global contract is deployed under certain account id it cannot be removed , only updated. So this code path indicates inconsistent storage, so use `RuntimeError::StorageError` instead of introducing a separate error kind here.